File: backend/src/main/java/com/roadmap/Roadmap/controllers/QuizData.java
package com.roadmap.Roadmap.controllers;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.CrossOrigin;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestParam;

import com.roadmap.Roadmap.models.Quiz.Question;
import com.roadmap.Roadmap.models.Quiz;
import com.roadmap.Roadmap.repositories.QuizRepo;

public class QuizData {
    
    @Autowired
    QuizRepo Quizzes;

    @GetMapping("/getQuiz")
    @CrossOrigin(origins = "*")
    public Quiz getQuizDataById(@RequestParam String id){
        return Quizzes.findById(id).orElse(null);
    }

    @PostMapping("/addNewQuiz")
    @CrossOrigin(origins = "*")
    public Quiz addNewQuiz(@RequestBody Quiz details){
        return Quizzes.save(details);
    }

    @PatchMapping("/updateQuiz")
    @CrossOrigin(origins = "*")
    public Quiz updateQuiz(@RequestBody Quiz details){
        return Quizzes.save(details);
    }

    @PatchMapping("/addNewQuestion")
    @CrossOrigin(origins = "*")
    public Quiz addNewQuestion(@RequestParam String id, @RequestBody Question question){
        Quiz quiz=Quizzes.findById(id).orElse(null);
        quiz.questions.add(question);
        return quiz;
    }

    @PatchMapping("/deleteQuestion")
    @CrossOrigin(origins = "*")
    public Quiz deleteQuestion(@RequestParam String id, @RequestBody Integer questionNo){
        Quiz quiz=Quizzes.findById(id).orElse(null);
        quiz.questions.remove(questionNo-1);
        return quiz;
    }

    public class UpdateQuestionBody{
        Integer questionNo;
        Question question;
    }
    @PatchMapping("/updateQuestion")
    @CrossOrigin(origins = "*")
    public Quiz updateQuestion(@RequestParam String id, @RequestBody UpdateQuestionBody body){
        Quiz quiz=Quizzes.findById(id).orElse(null);
        quiz.questions.setElementAt(body.question, body.questionNo);
        return quiz;
    }
}

File: backend/controllers/QuizData.js
const express = require("express");
const router = express.Router();
const Quiz = require("../models/Quiz");

// GET quiz data by ID
router.get("/getQuiz", async (req, res) => {
  try {
    const quiz = await Quiz.findById(req.query.id);
    res.json(quiz);
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
});

// POST add new quiz
router.post("/addNewQuiz", async (req, res) => {
  try {
    const quiz = await Quiz.save(req.body);
    res.json(quiz);
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
});

// PATCH update quiz
router.patch("/updateQuiz", async (req, res) => {
  try {
    const quiz = await Quiz.save(req.body);
    res.json(quiz);
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
});

// PATCH add new question
router.patch("/addNewQuestion", async (req, res) => {
  try {
    const quiz = await Quiz.findById(req.query.id);
    quiz.questions.push(req.body);
    const updatedQuiz = await Quiz.save(quiz);
    res.json(updatedQuiz);
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
});

// PATCH delete question
router.patch("/deleteQuestion", async (req, res) => {
  try {
    const quiz = await Quiz.findById(req.query.id);
    quiz.questions.splice(req.body.questionNo - 1, 1);
    const updatedQuiz = await Quiz.save(quiz);
    res.json(updatedQuiz);
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
});

// PATCH update question
router.patch("/updateQuestion", async (req, res) => {
  try {
    const quiz = await Quiz.findById(req.query.id);
    quiz.questions[req.body.questionNo - 1] = req.body.question;
    const updatedQuiz = await Quiz.save(quiz);
    res.json(updatedQuiz);
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
});

module.exports = router;
